src/yaml/YamlSchema.ts

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ */
+import * as vscode from 'vscode';
+import * as semver from 'semver';
+import { VSCODE_YAML_EXTENSION_ID, MICROPROFILE_SCHEMA, MICROPROFILE_SCHEMA_PREFIX } from "./YamlConstants";
+import { Uri } from 'vscode';
+import { LanguageClient } from 'vscode-languageclient';
+
+// The function signature exposed by vscode-yaml:
+// 1. the requestSchema api will be called by vscode-yaml extension to decide whether the schema can be handled by this
+// contributor, if it returns undefined, means it doesn't support this yaml file, vscode-yaml will ask other contributors
+// 2. the requestSchemaContent api  will give the parameter uri returned by the first api, and ask for the json content(after stringify) of
+// the schema
+declare type YamlSchemaContributor = (schema: string,
+  requestSchema: (resource: string) => string | undefined,
+  requestSchemaContent: (uri: string) => Promise<string>) => void;
+
+interface JsonSchemaForProjectInfo {
+  projectURI: string;
+  jsonSchema: string;
+}
+
+interface MicroProfilePropertiesChangeEvent {
+  projectURIs: Array<String>;
+}
+
+// Yaml Schema cache which caches YAML Schema (in JSON format) for application.yaml files
+// which belong to a MicroProfile project.
+export class YamlSchemaCache {
+  private readonly cache = new Map<string /* application yaml URI */, JsonSchemaForProjectInfo>();
+  private _languageClient: LanguageClient;
+  public set languageClient(languageClient: LanguageClient) {
+    this._languageClient = languageClient;
+  }
+
+  /**
+   * Returns the JSON Schema to use for the  given application yaml file URI.
+   *
+   * @param applicationYamlUri the application yaml file URI
+   */
+  public async getSchema(applicationYamlUri: string): Promise<string | undefined> {
+    const yamlSchema = this.cache.get(applicationYamlUri);
+    if (yamlSchema) {
+      return yamlSchema.jsonSchema;
+    }
+    if (!this._languageClient) {
+      return undefined;
+    }
+    const params = {
+      uri: applicationYamlUri,
+      scopes: [1, 2]
+    };
+    return this._languageClient.sendRequest('microprofile/jsonSchemaForProjectInfo', params)
+      .then((result: JsonSchemaForProjectInfo) => {
+        const jsonSchema = result.jsonSchema;
+        this.cache.set(applicationYamlUri, result);
+        return jsonSchema;
+      }, (err) => {
+        console.error("Error while consumming 'microprofile/jsonSchemaForProjectInfo' request: " + err.message);
+      });
+  }
+
+  /**
+   *  Evict the cache according to the classpath changed of the given project URIs
+   *
+   * @param event the properties change event
+   */
+  public evict(event: MicroProfilePropertiesChangeEvent) {
+    // 1. collect all application.yaml uri which belong to the project URIs
+    const uriToDelete: Array<string> = [];
+    for (const entry of this.cache.entries()) {
+      if (event.projectURIs.includes(entry[1].projectURI)) {
+        uriToDelete.push(entry[0]);
+      }
+    }
```

Comment:
I would consider using [`Map.forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/forEach) in this case: 

`this.cache.entries.forEach((value: JsonSchemaForProjectInfo, key: string) => {})`

Reply:
I'm sorry, I don't understand your suggestion?  The idea is to fill uriToDelete, could you explain more your suggestion please. Thanks!

Reply:
Sorry, what I meant was to use `forEach()` instead of `for (const entry of this.cache.entries())`.

Like this:
```
this.cache.forEach((jsonSchema: JsonSchemaForProjectInfo, uri: string) => {
    if (event.projectURIs.includes(jsonSchema.projectURI)) {
        uriToDelete.push(uri);
    }
});
```

Reply:
Ok thanks for the explanation, done.